crates/sui-types/src/move_package.rs

```diff
@@ -320,3 +433,104 @@ where
+    if !immediate_dependencies.is_empty() {
+        // TODO ExecutionError
+        panic!("Dependency not found in linkage table");
+    }
+
+    // (2) Every dependency's linkage table is superseded by this linkage table
+    for dep_linkage_table in dep_linkage_tables {
+        for (original_id, dep_info) in dep_linkage_table {
+            let Some(our_info) = linkage_table.get(original_id) else {
+                // TODO ExecutionError
+                panic!("Transitive dependency not found in linkage table");
+            };
+
+            if our_info.upgraded_version < dep_info.upgraded_version {
+                // TODO ExecutionError
+                panic!("Downgrade in linkage table");
+            }
+        }
+    }
+
+    Ok(linkage_table)
+}
+
+fn build_initial_type_origin_table(modules: &[CompiledModule]) -> BTreeMap<ModuleStruct, ObjectID> {
+    BTreeMap::from_iter(modules.iter().flat_map(|m| {
+        m.struct_defs().iter().map(|struct_def| {
+            let struct_handle = m.struct_handle_at(struct_def.struct_handle);
+            let module_name = m.name().to_string();
+            let struct_name = m.identifier_at(struct_handle.name).to_string();
+            let id: ObjectID = (*m.self_id().address()).into();
+            (
+                ModuleStruct {
+                    module_name,
+                    struct_name,
+                },
+                id,
+            )
+        })
+    }))
+}
+
+fn build_upgraded_type_origin_table(
+    predecessor: &MovePackage,
+    modules: &[CompiledModule],
+) -> BTreeMap<ModuleStruct, ObjectID> {
+    let mut new_table = BTreeMap::new();
+    // to double-check that the compatibility check for upgraded package was successful
+    let mut all_new_types = BTreeSet::new();
+    for m in modules {
+        for struct_def in m.struct_defs() {
+            let struct_handle = m.struct_handle_at(struct_def.struct_handle);
+            let module_name = m.name().to_string();
+            let struct_name = m.identifier_at(struct_handle.name).to_string();
+            let mod_struct = ModuleStruct {
+                module_name,
+                struct_name,
+            };
+            all_new_types.insert(mod_struct.clone());
+            // only insert types that are not in the original table as only these should be
+            // marked as originating from the current package
+            if predecessor.type_origin_table.contains_key(&mod_struct) {
+                continue;
+            }
+            let id: ObjectID = (*m.self_id().address()).into();
+            new_table.insert(mod_struct, id);
+        }
+    }
+    if !BTreeSet::from_iter(predecessor.type_origin_table.keys().cloned()).is_subset(&all_new_types)
+    {
+        // panic as it should be caught by compatibility check
+        panic!("Earlier package version contains a type missing from the upgraded one");
+    }
+    new_table.extend(predecessor.type_origin_table.clone().into_iter());
+    new_table
```

Comment:
Slightly "neater" (?) construction below (rather than tracking all structs, track existing structs by removing them from a set and checking that the set you are removing them from is empty at the end).

Also here we shouldn't panic, we should produce an invariant violation of some kind, because the issue is local/contained, we don't need to bring the whole network down because of it:

```suggestion
    let mut new_table = BTreeMap::new();
    let mut existing_table = predecessor.type_origin_table.clone();
    for m in modules {
        for struct_def in m.struct_defs() {
            let struct_handle = m.struct_handle_at(struct_def.struct_handle);
            let module_name = m.name().to_string();
            let struct_name = m.identifier_at(struct_handle.name).to_string();
            let mod_struct = ModuleStruct {
                module_name,
                struct_name,
            };
            
            let id = existing_table.remove(&mod_struct).unwrap_or_else(|| {
                (*m.self_id().address()).into()
            });
            
            new_table.insert(mod_struct, id);
        }
    }
    
    if !existing_table.is_empty() {
        // TODO: Invariant violation to replace this panic.
        panic!("Earlier package version contains a type missing from the upgraded one");
    }
    
    new_table
```